pkg/apis/eventing/v1alpha1/cluster_provisioner_types.go

```diff
@@ -0,0 +1,117 @@
+type ClusterProvisioner struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ClusterProvisionerSpec   `json:"spec"`
+	Status ClusterProvisionerStatus `json:"status"`
+}
+
+// Check that ClusterProvisioner can be validated and can be defaulted.
+var _ apis.Validatable = (*ClusterProvisioner)(nil)
+var _ apis.Defaultable = (*ClusterProvisioner)(nil)
+var _ runtime.Object = (*ClusterProvisioner)(nil)
+var _ webhook.GenericCRD = (*ClusterProvisioner)(nil)
+
+// ClusterProvisionerSpec is the spec for a ClusterProvisioner resource.
+type ClusterProvisionerSpec struct {
+	// TODO: Generation does not work correctly with CRD. They are scrubbed
+	// by the APIserver (https://github.com/kubernetes/kubernetes/issues/58778)
+	// So, we add Generation here. Once that gets fixed, remove this and use
+	// ObjectMeta.Generation instead.
+	// +optional
+	Generation int64 `json:"generation,omitempty"`
+
+	// Type is the type of the resource to be provisioned.
+	// +required
+	Type runtime.TypeMeta `json:"type"`
+}
+
+type ClusterProvisionerConditionType string
+
+const (
+	// ClusterProvisionerConditionReady specifies that the resource is ready.
+	ClusterProvisionerConditionReady ClusterProvisionerConditionType = "Ready"
+)
+
+// ClusterProvisionerConditionStatus describes the state of this resource at a point in time.
+type ClusterProvisionerConditionStatus struct {
+	// Type of condition.
+	// +required
+	Type ClusterProvisionerConditionType `json:"type"`
+
+	// Status of the condition, one of True, False, Unknown.
+	// +required
+	Status corev1.ConditionStatus `json:"status"`
+
+	// LastTransitionTime is the last time the condition transitioned from one status to another.
+	// We use VolatileTime in place of metav1.Time to exclude this from creating equality.Semantic
+	// differences (all other things held constant).
+	// +optional
+	LastTransitionTime apis.VolatileTime `json:"lastTransitionTime,omitempty" description:"last time the condition transit from one status to another"`
+
+	// The reason for the condition's last transition.
+	// +optional
+	Reason string `json:"reason,omitempty"`
+
+	// A human readable message indicating details about the transition.
+	// +optional
+	Message string `json:"message,omitempty"`
+}
+
+// ClusterProvisionerStatus is the status for a ClusterProvisioner resource
+type ClusterProvisionerStatus struct {
+	// Conditions holds the state of a cluster provisioner at a point in time.
+	Conditions []ClusterProvisionerConditionStatus `json:"conditions,omitempty"`
+
+	// ObservedGeneration is the 'Generation' of the ClusterProvisioner that
+	// was last reconciled by the controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+}
+
+// GetSpecJSON returns spec as json
+func (p *ClusterProvisioner) GetSpecJSON() ([]byte, error) {
+	return json.Marshal(p.Spec)
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```

Comment:
Since this is cluster scoped, I think we need`// +genclient:nonNamespaced` here.

Reply:
Looking at the generated code for ClusterProvisionerLister, it looks like wey may not need to tell it this si `nonNamespaced`.